README.md
# rlecuyer

[![R build status](https://github.com/hanase/rlecuyer/workflows/R-CMD-check/badge.svg?branch=main)](https://github.com/hanase/rlecuyer/actions?workflow=R-CMD-check)

R interface to P. L'Ecuyer's Random Number Generator with multiple streams. The package can be installed from CRAN.



ChangeLog
v0.3-6/7 - 03/23/2023:
 - fixed the definitions of C functions with no arguments.
 - help page for the package added as well as more examples.

v0.3-5 - 11/19/2019:
 - Wei-Chen Chen fixed a NOTE regarding registering dynamic C calls.

v0.3-4.1 - 11/17/2019:
 - Changes made by Kurt Hornik (fixes return values for functions called by .Call)
 
v0.3-4 - 09/17/2015:
 - Pierre L'Ecuyer added to Authors
 - Fixed wrong "UNPROTECT"ing objects

v0.3-3 - 02/01/2012:
 - .First.Lib replaced by .onLoad, .Last.Lib by .onUnload

v0.3-2 - 01/25/2012: 
 - NAMESPACE added
 - In the C code, functions printf and exit replaced by Rprintf and error, respectively.
 - Minor cosmetic changes.

04/29/2004: Initial version


src/zzz.c
#include <R.h>
#include <Rdefines.h>
#include <Rinternals.h>
#include <R_ext/Rdynload.h>

SEXP r_advance_state (SEXP sexp_e, SEXP sexp_c,
                      SEXP sexp_streamCg,SEXP sexp_streamBg,
                      SEXP sexp_streamIg,
                      SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
                      SEXP sexp_streamName);
SEXP r_create_current_stream (void);
SEXP r_create_stream (SEXP sexp_name);
SEXP r_get_current_stream(void);
SEXP r_randU01 (SEXP sexp_streamCg,SEXP sexp_streamBg,SEXP sexp_streamIg,
                SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
                SEXP sexp_streamName);
SEXP r_remove_current_stream (void);
SEXP r_reset_next_substream (SEXP sexp_streamCg,SEXP sexp_streamBg,
                             SEXP sexp_streamIg,
                             SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
                             SEXP sexp_streamName);
SEXP r_set_current_stream (SEXP sexp_streamCg,SEXP sexp_streamBg,
                           SEXP sexp_streamIg,
                           SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
                           SEXP sexp_streamName);
SEXP r_set_package_seed (SEXP sexp_seed);
SEXP r_set_stream_seed (SEXP sexp_seed, SEXP sexp_streamCg,SEXP sexp_streamBg,
                        SEXP sexp_streamIg,
                        SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
                        SEXP sexp_streamName);

static const R_CallMethodDef callMethods[] = {
	{"r_advance_state", (DL_FUNC) &r_advance_state, 8},
	{"r_create_current_stream", (DL_FUNC) &r_create_current_stream, 0},
	{"r_create_stream", (DL_FUNC) &r_create_stream, 1},
	{"r_get_current_stream", (DL_FUNC) &r_get_current_stream, 0},
	{"r_randU01", (DL_FUNC) &r_randU01, 6},
	{"r_remove_current_stream", (DL_FUNC) &r_remove_current_stream, 0},
	{"r_reset_next_substream", (DL_FUNC) &r_reset_next_substream, 6},
	{"r_set_current_stream", (DL_FUNC) &r_set_current_stream, 6},
	{"r_set_package_seed", (DL_FUNC) &r_set_package_seed, 1},
	{"r_set_stream_seed", (DL_FUNC) &r_set_stream_seed, 7},

	/* Finish R_CallMethodDef. */
	{NULL, NULL, 0}
};
/* End of the callMethods[]. */

void R_init_rlecuyer(DllInfo *info){
	R_registerRoutines(info, NULL, callMethods, NULL, NULL);
	R_useDynamicSymbols(info, TRUE);
} /* End of R_init_rlecuyer(). */


src/RngWrap.c
#include <R.h>
#include <Rdefines.h>
#include <R_ext/Random.h>
#include "RngStream.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static RngStream current_stream;
static double rn = 0.0;

SEXP r_create_current_stream(void)
{
 current_stream = (RngStream) malloc (sizeof (struct RngStream_InfoState));

 if (current_stream == NULL) {
    error("r_create_current_stream: No more memory\n\n");
  }
 current_stream->name="";
 return R_NilValue;
}

SEXP r_remove_current_stream(void)
{
  free(current_stream);
  /*    RngStream_DeleteStream(&current_stream);*/

  return R_NilValue;
}

SEXP r_create_stream (SEXP sexp_name)
{
  RngStream newstream;
  SEXP sexp_newstream;
  int i;

  const char *name = CHAR(STRING_ELT(sexp_name,0));
  newstream =  RngStream_CreateStream(name);

  PROTECT(sexp_newstream = allocVector(REALSXP, 20));

  for (i = 0; i < 6; ++i) {
    REAL(sexp_newstream)[i] = (double) newstream->Cg[i];
    REAL(sexp_newstream)[i+6] = (double) newstream->Bg[i];
    REAL(sexp_newstream)[i+12] = (double) newstream->Ig[i];
  }
  REAL(sexp_newstream)[18] = (int) newstream->Anti;
  REAL(sexp_newstream)[19] = (int) newstream->IncPrec;
  UNPROTECT(1);

  RngStream_DeleteStream(&newstream);
  return(sexp_newstream);
}

SEXP r_set_package_seed(SEXP sexp_seed)
{
  int i;
  unsigned long seed[6];

  for (i = 0; i < 6; ++i) {
    seed[i]= (unsigned long) REAL(sexp_seed)[i];
  }
  RngStream_SetPackageSeed( (unsigned long *) seed);
  return R_NilValue;
}


SEXP r_set_stream_seed(SEXP sexp_seed, SEXP sexp_streamCg,SEXP sexp_streamBg,
		       SEXP sexp_streamIg,
		       SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
		       SEXP sexp_streamName)
{
  SEXP sexp_stream;
  RngStream stream;
  unsigned long seed[6];
  int i;
  const char *name;
  size_t len;

  stream = (RngStream) malloc (sizeof (struct RngStream_InfoState));
  if (stream == NULL) {
    error("r_set_stream_seed: No more memory\n\n");
  }
 
  for (i=0; i<6; i++) {
    stream->Cg[i] = REAL(sexp_streamCg)[i];
    stream->Bg[i] = REAL(sexp_streamBg)[i];
    stream->Ig[i] = REAL(sexp_streamIg)[i];
  }
  stream->Anti = INTEGER(sexp_streamAnti)[0];
  stream->IncPrec = INTEGER(sexp_streamIncPrec)[0];
  name = CHAR(STRING_ELT(sexp_streamName,0));
  len = strlen(name);
  stream->name = malloc(len+1);
  if (stream->name == NULL) {
      free(stream);
      error("r_set_stream_seed: No more memory\n\n");
    }
  strncpy(stream->name, name, len+1);

  for (i = 0; i < 6; ++i) {
    seed[i]= (unsigned long) REAL (sexp_seed)[i];
  }

  RngStream_SetSeed(stream, (unsigned long *) seed);
  PROTECT(sexp_stream = allocVector(REALSXP, 20));

  for (i = 0; i < 6; ++i) {
    REAL(sexp_stream)[i] = (double) stream->Cg[i];
    REAL(sexp_stream)[i+6] = (double) stream->Bg[i];
    REAL(sexp_stream)[i+12] = (double) stream->Ig[i];
  }
  REAL(sexp_stream)[18] = (int) stream->Anti;
  REAL(sexp_stream)[19] = (int) stream->IncPrec;
  UNPROTECT(1);

  free(stream);
  return sexp_stream;
}


SEXP r_randU01 (SEXP sexp_streamCg,SEXP sexp_streamBg,SEXP sexp_streamIg,
		SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
		SEXP sexp_streamName)
{
  SEXP sexp_result;
  RngStream stream;
  int i;
  double rn = 0.0;
  const char *name;
  size_t len;

  stream = (RngStream) malloc (sizeof (struct RngStream_InfoState));
  if (stream == NULL) {
    error("r_randU01: No more memory\n\n");
  }
 
  for (i=0; i<6; i++) {
    stream->Cg[i] = REAL(sexp_streamCg)[i];
    stream->Bg[i] = REAL(sexp_streamBg)[i];
    stream->Ig[i] = REAL(sexp_streamIg)[i];
  }
  stream->Anti = INTEGER(sexp_streamAnti)[0];
  stream->IncPrec = INTEGER(sexp_streamIncPrec)[0];
  name = CHAR(STRING_ELT(sexp_streamName,0));
  len = strlen(name);
  stream->name = malloc(len+1);
  if (stream->name == NULL) {
    free(stream);
    error("r_randU01: No more memory\n\n");
  }
  strncpy(stream->name, name, len+1);

  rn = RngStream_RandU01(stream);

  PROTECT(sexp_result = allocVector(REALSXP, 21));

  for (i = 0; i < 6; ++i) {
    REAL(sexp_result)[i] = (double) stream->Cg[i];
    REAL(sexp_result)[i+6] = (double) stream->Bg[i];
    REAL(sexp_result)[i+12] = (double) stream->Ig[i];
  }
  REAL(sexp_result)[18] = (int) stream->Anti;
  REAL(sexp_result)[19] = (int) stream->IncPrec;
  REAL(sexp_result)[20] = (double) rn;
  UNPROTECT(1);

  free(stream);
  return sexp_result;
}

SEXP r_reset_next_substream (SEXP sexp_streamCg,SEXP sexp_streamBg,
			     SEXP sexp_streamIg,
			     SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
			     SEXP sexp_streamName)
{
  SEXP sexp_stream;
  RngStream stream;
  int i;
  const char *name;
  size_t len;

  stream = (RngStream) malloc (sizeof (struct RngStream_InfoState));
  if (stream == NULL) {
    error("r_reset_next_substream: No more memory\n\n");
  }
 
  for (i=0; i<6; i++) {
    stream->Cg[i] = REAL(sexp_streamCg)[i];
    stream->Bg[i] = REAL(sexp_streamBg)[i];
    stream->Ig[i] = REAL(sexp_streamIg)[i];
  }
  stream->Anti = INTEGER(sexp_streamAnti)[0];
  stream->IncPrec = INTEGER(sexp_streamIncPrec)[0];
  name = CHAR(STRING_ELT(sexp_streamName,0));
  len = strlen(name);
  stream->name = malloc(len+1);
  if (stream->name == NULL) {
    free(stream);
    error("r_set_stream_seed: No more memory\n\n");
  }
  strncpy(stream->name, name, len+1);

  RngStream_ResetNextSubstream (stream);

  PROTECT(sexp_stream = allocVector(REALSXP, 20));

  for (i = 0; i < 6; ++i) {
    REAL(sexp_stream)[i] = (double) stream->Cg[i];
    REAL(sexp_stream)[i+6] = (double) stream->Bg[i];
    REAL(sexp_stream)[i+12] = (double) stream->Ig[i];
  }
  REAL(sexp_stream)[18] = (int) stream->Anti;
  REAL(sexp_stream)[19] = (int) stream->IncPrec;
  UNPROTECT(1);

  free(stream);
  return sexp_stream;
}

SEXP r_advance_state (SEXP sexp_e, SEXP sexp_c, 
		      SEXP sexp_streamCg,SEXP sexp_streamBg,
		      SEXP sexp_streamIg,
		      SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
		      SEXP sexp_streamName)
{
  SEXP sexp_stream;
  RngStream stream;
  int i;
  long e, c;
  const char *name;
  size_t len;

  stream = (RngStream) malloc (sizeof (struct RngStream_InfoState));
  if (stream == NULL) {
    error("r_advance_state: No more memory\n\n");
  }
 
  e = REAL(sexp_e)[0];
  c = REAL(sexp_c)[0];
  for (i=0; i<6; i++) {
    stream->Cg[i] = REAL(sexp_streamCg)[i];
    stream->Bg[i] = REAL(sexp_streamBg)[i];
    stream->Ig[i] = REAL(sexp_streamIg)[i];
  }
  stream->Anti = INTEGER(sexp_streamAnti)[0];
  stream->IncPrec = INTEGER(sexp_streamIncPrec)[0];
  name = CHAR(STRING_ELT(sexp_streamName,0));
  len = strlen(name);
  stream->name = malloc(len+1);
  if (stream->name == NULL) {
    free(stream);
    error("r_set_stream_seed: No more memory\n\n");
  }
  strncpy(stream->name, name, len+1);

  RngStream_AdvanceState (stream, e, c);

  PROTECT(sexp_stream = allocVector(REALSXP, 20));

  for (i = 0; i < 6; ++i) {
    REAL(sexp_stream)[i] = (double) stream->Cg[i];
    REAL(sexp_stream)[i+6] = (double) stream->Bg[i];
    REAL(sexp_stream)[i+12] = (double) stream->Ig[i];
  }
  REAL(sexp_stream)[18] = (int) stream->Anti;
  REAL(sexp_stream)[19] = (int) stream->IncPrec;
  UNPROTECT(1);

  free(stream);
  return sexp_stream;
}

SEXP r_set_current_stream (SEXP sexp_streamCg,SEXP sexp_streamBg,
		       SEXP sexp_streamIg,
		       SEXP sexp_streamAnti,SEXP sexp_streamIncPrec,
		       SEXP sexp_streamName)
{
  int i;
  const char *name;
  size_t len;

  for (i=0; i<6; i++) {
    current_stream->Cg[i] = REAL(sexp_streamCg)[i];
    current_stream->Bg[i] = REAL(sexp_streamBg)[i];
    current_stream->Ig[i] = REAL(sexp_streamIg)[i];
  }

  current_stream->Anti = INTEGER(sexp_streamAnti)[0];
  current_stream->IncPrec = INTEGER(sexp_streamIncPrec)[0];

  name = CHAR(STRING_ELT(sexp_streamName,0));
  len = strlen(name);
  current_stream->name = malloc(len+1);
  if (current_stream->name == NULL) {
    free(current_stream);
    error("r_set_current_stream: No more memory\n\n");
  }
  strncpy(current_stream->name, name, len+1);

  return R_NilValue;
}

SEXP r_get_current_stream(void)
{
  SEXP sexp_stream,dummy,dummys;
  int i;

  PROTECT(dummy = allocVector(REALSXP, 20));
  for (i = 0; i < 6; ++i) {
    REAL(dummy)[i] = (double) current_stream->Cg[i];
    REAL(dummy)[i+6] = (double) current_stream->Bg[i];
    REAL(dummy)[i+12] = (double) current_stream->Ig[i];
  }
  REAL(dummy)[18] = (int) current_stream->Anti;
  REAL(dummy)[19] = (int) current_stream->IncPrec;

  PROTECT(dummys = allocVector(STRSXP, 1));
  SET_STRING_ELT(dummys,0, mkChar(current_stream->name));

  PROTECT(sexp_stream = allocVector(VECSXP, 2));
  SET_VECTOR_ELT(sexp_stream,0,dummy);
  SET_VECTOR_ELT(sexp_stream,1, dummys);
  UNPROTECT(3);
  return sexp_stream;
}

double *user_unif_rand (void)
{
  if (!current_stream) {
    Rprintf("No stream created yet!");
    return NULL;
  }
  rn = RngStream_RandU01(current_stream);
  return &rn;
}


R/rlecuyer.R
.lec.init <- function () {
  if(exists(".lec.Random.seed.table", envir=.GlobalEnv))
    rm(".lec.Random.seed.table", envir=.GlobalEnv)
  pos <- 1
  assign(".lec.Random.seed.table",list(Cg=matrix(0,nrow=0,ncol=6),
                                       Bg=matrix(0,nrow=0,ncol=6),
                                       Ig=matrix(0,nrow=0,ncol=6),
                                       AIP=matrix(0,nrow=0,ncol=2),
                                       name=c()), envir=as.environment(pos))
  .Call("r_create_current_stream",PACKAGE="rlecuyer")
  return(1)
}

.lec.exit <- function () {
  if(exists(".lec.Random.seed.table", envir=.GlobalEnv))
    rm(".lec.Random.seed.table", envir=.GlobalEnv)
  .Call("r_remove_current_stream",PACKAGE="rlecuyer")
  return(1)
}

.lec.which.stream <- function (name) {
  if(!exists(".lec.Random.seed.table", envir=.GlobalEnv))
    stop("No stream created yet");
  which <- which(.lec.Random.seed.table$name == name)
  if (length(which) <= 0) return (0)
  return(which[1])
}

.lec.GetStateList <- function (name) {
  which <- .lec.which.stream(name)
  if (which == 0) 
    stop(paste("No stream ",name," exists."));
  return (list(Cg=.lec.Random.seed.table$Cg[which,],Bg=.lec.Random.seed.table$Bg[which,],Ig=.lec.Random.seed.table$Ig[which,],Anti=.lec.Random.seed.table$AIP[which,1],IncPrec=.lec.Random.seed.table$AIP[which,2],name=.lec.Random.seed.table$name[which]))
}

.lec.SaveStateList <- function (stream) {
  which <- .lec.which.stream(stream$name)
  if (which == 0) 
    stop(paste("No stream ",stream$name," exists."));
  
  .lec.Random.seed.table$Cg[which,] <<- stream$Cg
  .lec.Random.seed.table$Bg[which,] <<- stream$Bg
  .lec.Random.seed.table$Ig[which,] <<- stream$Ig
  .lec.Random.seed.table$AIP[which,] <<- c(stream$Anti,stream$IncPrec)
  .lec.Random.seed.table$name[which] <<- stream$name
  return(which)
}

.lec.DoCreateStream <- function (name) {

  state <- .Call("r_create_stream", as.character(name),PACKAGE="rlecuyer")
  stream <- list(Cg=state[1:6],Bg=state[7:12],Ig=state[13:18],
                 Anti=as.integer(state[19]), IncPrec=as.integer(state[20]),
                 name=name)
  return(stream)
}

.lec.CreateStream <- function (names) {
  if(!exists(".lec.Random.seed.table", envir=.GlobalEnv))
    .lec.init()

  ln <- length(names)
  if (ln == 0)
    stop("Stream name is missing.")
  
  for (i in 1:ln) {
    name <- names[i]
  
    g <- .lec.DoCreateStream(name)
    which <- .lec.which.stream(name)
  
    if (which == 0) { # not in the table
      .lec.Random.seed.table$Cg <<- rbind(.lec.Random.seed.table$Cg,g$Cg[1:6])
      .lec.Random.seed.table$Bg <<- rbind(.lec.Random.seed.table$Bg,g$Bg)
      .lec.Random.seed.table$Ig <<- rbind(.lec.Random.seed.table$Ig,g$Ig)
      .lec.Random.seed.table$AIP <<- rbind(.lec.Random.seed.table$AIP,
                                           c(g$Anti, g$IncPrec))
      .lec.Random.seed.table$name <<- c(.lec.Random.seed.table$name, g$name)
    } else { # entry found in table
      warning("Name ", name, " already in use. Its state is overwritten.", immediate. = TRUE)
      .lec.SaveStateList(g)
    }
  }
}

.lec.WriteState <- function (names) {
  ln <- length(names)
  if (ln == 0)
    stop("Stream name is missing.")
  
  for (i in 1:ln) {
	name <- names[i]
  	g <- .lec.GetStateList(name)
  	cat("\nThe current state of the Rngstream ",name,":\n")
  	cat("Cg = ", g$Cg)
	}
}

.lec.WriteStateFull <- function (names) {
  ln <- length(names)
  if (ln == 0)
    stop("Stream name is missing.")
  
  for (i in 1:ln) {
	name <- names[i]
  	g <- .lec.GetStateList(name)
  	cat("\nThe Rngstream ",name,":\n")
  	cat("Anti = ",g$Anti)
  	cat("IncPrec = ",g$IncPrec, "\n")
  	cat("Ig = ",g$Ig, "\n")
  	cat("Bg = ",g$Bg, "\n")
  	cat("Cg = ",g$Cg, "\n")
	}
}

.lec.CheckSeed <- function(seed) {
  ll<-length(seed)
  if (ll < 6) {
    tmpseed<-12345
    seed<-c(seed,rep(tmpseed,6-ll))
  }
  if (ll > 6) {
  	seed<-seed[1:6]
    warning("Seed may not exceed the length of 6. Truncated to ", 
    		paste(seed, collapse=', '))
  }
  return(seed)
}

.lec.SetPackageSeed <- function(seed=rep(12345,6)) {
  if(!exists(".lec.Random.seed.table", envir=.GlobalEnv))
    .lec.init()
  seed <- .lec.CheckSeed(seed)
  .Call ("r_set_package_seed", as.double(seed), PACKAGE="rlecuyer")
  return(seed)
}

.lec.SetSeed <- function(name, seed=rep(12345,6)) {
  which <- .lec.which.stream(name)
  if (which == 0) 
    stop(paste("No stream ",name," exists."));
  
  seed <- .lec.CheckSeed(seed)
  g <- .lec.GetStateList(name)

  state <- .Call ("r_set_stream_seed", as.double(seed), as.double(g$Cg),
                 as.double(g$Bg),as.double(g$Ig),as.integer(g$Anti),
                 as.integer(g$IncPrec),as.character(g$name),
                 PACKAGE="rlecuyer")
  g<-list(Cg=state[1:6],Bg=state[7:12],Ig=state[13:18],
          Anti=as.integer(state[19]), IncPrec=as.integer(state[20]),
          name=name)

  .lec.SaveStateList(g)
  return(seed)
}

.lec.GetState <- function(name){
  g <- .lec.GetStateList(name)
  return(g$Cg)
}

.lec.Getuniform <- function(stream) {
  number<-.Call("r_randU01",as.double(stream$Cg),as.double(stream$Bg),
                as.double(stream$Ig),as.integer(stream$Anti),
                as.integer(stream$IncPrec),as.character(stream$name),
                PACKAGE="rlecuyer")
  return(number)
}

.lec.uniform <- function(name,n=1) {  
  g <- .lec.GetStateList(name)
  sample<-rep(0,n)
  for (i in 1:n) {
    res<-.lec.Getuniform(g)
    sample[i] <- res[[length(res)]]
    g <- list(Cg=res[1:6],Bg=res[7:12],Ig=res[13:18],
              Anti=res[19],IncPrec=res[20],name=name)
  }
  .lec.SaveStateList(g)
  return(sample)
}


.lec.uniform.int <- function (name, n=1, a=0,b=10) {
  rn <- .lec.uniform (name,n)
  rn <- as.integer(a + (b-a+1)*rn)
  return(rn)
}

.lec.DeleteStream <- function (names) {
  ln <- length(names)
  for (i in 1:ln) {
    name <- names[i]
    which <- .lec.which.stream(name)
    if (which == 0) {# not in the table
      warning("Stream ",name," does not exist.", immediate. = TRUE)
    } else { # found
      if (length(.lec.Random.seed.table$name) > 1) {
        .lec.Random.seed.table$Cg<<-.lec.Random.seed.table$Cg[-which,]
        .lec.Random.seed.table$Bg<<-.lec.Random.seed.table$Bg[-which,]
        .lec.Random.seed.table$Ig<<-.lec.Random.seed.table$Ig[-which,]
        .lec.Random.seed.table$AIP<<-.lec.Random.seed.table$AIP[-which,]
        .lec.Random.seed.table$name <<- .lec.Random.seed.table$name[-which]
      } else  {# table empty
      	pos <- 1
      	assign(".lec.Random.seed.table",list(Cg=matrix(0,nrow=0,ncol=6),
                                           Bg=matrix(0,nrow=0,ncol=6),
                                           Ig=matrix(0,nrow=0,ncol=6),
                                           AIP=matrix(0,nrow=0,ncol=2),
                                           name=c()), envir=as.environment(pos))
      }
    }
  }
}

.lec.StreamExists <- function(name) {
  which <-  which(.lec.Random.seed.table$name == name)
  if (length(which) > 0) return(TRUE)
  return(FALSE)
}

.lec.ResetStartStream <- function(name) {
  which <- .lec.which.stream(name)
  if (which == 0) 
    stop(paste("No stream ",name," exists."));
  .lec.Random.seed.table$Cg[which,] <<- .lec.Random.seed.table$Bg[which,] <<- .lec.Random.seed.table$Ig[which,]
}

.lec.ResetStartSubstream <- function(name) {
  which <- .lec.which.stream(name)
  if (which == 0) 
    stop(paste("No stream ",name," exists."));
  .lec.Random.seed.table$Cg[which,] <<- .lec.Random.seed.table$Bg[which,]
}

.lec.ResetNextSubstream <- function(name) {
  g <- .lec.GetStateList(name)  
  state <- .Call("r_reset_next_substream", as.double(g$Cg),
                 as.double(g$Bg),as.double(g$Ig),as.integer(g$Anti),
                 as.integer(g$IncPrec),as.character(g$name),
                 PACKAGE="rlecuyer")
  g<-list(Cg=state[1:6],Bg=state[7:12],Ig=state[13:18],
                 Anti=as.integer(state[19]), IncPrec=as.integer(state[20]),
                 name=name)
  .lec.SaveStateList(g)
}

.lec.SetAntithetic <- function (name, anti=FALSE) {
  which <- .lec.which.stream(name)
  if (which == 0) 
    stop(paste("No stream ",name," exists."));
  .lec.Random.seed.table$AIP[which,1] <<- as.integer(anti)
}

.lec.IncreasedPrecis <- function (name, incp=FALSE) {
  which <- .lec.which.stream(name)
  if (which == 0) 
    stop(paste("No stream ",name," exists."));
  .lec.Random.seed.table$AIP[which,2] <<- as.integer(incp)
}

.lec.AdvanceState <- function (name, e=1, c=0) {
  if (!is.numeric(e) | !is.numeric(c))
    stop(".lec.AdvanceState: Arguments must be numeric.")
  
  g <- .lec.GetStateList(name)  
  state <- .Call("r_advance_state", as.double(e), as.double(c),
                 as.double(g$Cg),
                 as.double(g$Bg),as.double(g$Ig),as.integer(g$Anti),
                 as.integer(g$IncPrec),as.character(g$name),
                 PACKAGE="rlecuyer")
  g<-list(Cg=state[1:6],Bg=state[7:12],Ig=state[13:18],
                 Anti=as.integer(state[19]), IncPrec=as.integer(state[20]),
                 name=name)
  .lec.SaveStateList(g)
  }

.lec.GetStreams <- function() {
  if(!exists(".lec.Random.seed.table", envir=.GlobalEnv))
    return(NULL)
  return(.lec.Random.seed.table$name)
}

.lec.CurrentStream <- function(name) {
  old.kind<-RNGkind()
  if (exists (".Random.seed") && floor(.Random.seed[1]/100) == 5) {
        # Leftover from have used user defined rng and forgot
        # to switch back after last run
        set.seed (0)
    }
  RNGkind ("user")
  g <- .lec.GetStateList(name)
  .Call("r_set_current_stream",as.double(g$Cg),
         as.double(g$Bg),as.double(g$Ig),as.integer(g$Anti),
         as.integer(g$IncPrec),as.character(g$name),
         PACKAGE="rlecuyer" )
  return(old.kind)
}

.lec.CurrentStreamEnd <- function(kind.old = c("Mersenne-Twister", "Kinderman-Ramage")) {
   stream <- .Call("r_get_current_stream",PACKAGE="rlecuyer" )


   if (as.character(stream[[2]])=="") return(NULL) 
   g<-list(Cg=stream[[1]][1:6],Bg=stream[[1]][7:12],Ig=stream[[1]][13:18],
           Anti=as.integer(stream[[1]][19]),
           IncPrec=as.integer(stream[[1]][20]),
           name=as.character(stream[[2]]))
   .lec.SaveStateList(g)
    RNGkind (kind.old[1], kind.old[2])
   return(g$name)
}
